### DISCUSSION TITLE
Is it time to rewrite Node's internals in a memory safe modern language?

### DISCUSSION THREAD

**Discussion by jsoneaday** [nodejs/help#4592] [Ideas]
With the advent of projects like Bun and Deno, both of which use modern memory safe languages, is it time to rewrite Node internals too? Both of these runtimes seem to have benefited from their choice of using Zig and Rust. 
Although benchmarks are not the end all be all, at a rough level both projects seem to be generally more performant than Node. In addition the memory safety of these languages should lead to less memory related bugs in the longer term.
There may also be room for improvements in the core design of libuv including doing a total redesign.

**Comment by xTrayambak** [comment]
Zig isn't memory safe. It just makes allocations more explicit.
Rust can't prevent memory leaks on its own like a traditional RC can.

**Comment by kravetsone** [comment]
Why u need NodeJS rewrite? Just use what u want...

**Comment by marco-ippolito** [comment]
PR WELCOME 😄

**Reply by kontroinc** [reply]
lol
a brilliant point
